### DISCUSSION TITLE
What's the best way of keeping Miniforge up to date?

### DISCUSSION THREAD

**Discussion by gernophil** [conda-forge/miniforge#649] [Q&A] [answered]
I know this question hast been asked a lot and you are probably annoyed by it. But what is the best way of keeping Miniforge up to date?

The command I read mostly nowadays is:
```
mamba update --all
```

But there is also:
```
mamba update mamba
mamba update --all
```

Earlier there was:
```
(conda update conda)
conda update --all)
```

1. I guess the outcome of the `conda` command would be equivalent to the `mamba` command, am I correct?
2. But what is the use of updating `mamba` or `conda`, before updating `--all`?

Besides that there is also the `-u` flag of the installer:
```
-u           update an existing installation
```

3. What's the use of that and what are the differences in outcome of updating using the installer?

I always do a fresh reinstall after uninstalling once in a while, but that's always a little time consuming since I also have to do all the config stuff. This is of course doable, but it would be nice, if there was one official way of keeping conda up to date.

Also for this I have some questions:

4. What would be the difference in outcome of a fresh reinstall vs. the `-u` way vs. the `mamba update --all` way?
5. And what is the preferred way?

I also feel it would be great, if the one official way would be mentioned in the docs.

Thanks for elaborating :).

**Comment by hmaarrfk** [comment, marked as answer]
1. I guess the outcome of the conda command would be equivalent to the mamba command, am I correct?

conda and mamba are not 100% equivalent. Sometimes there are slight differences, but for `update` and `install` in 2024, they should be mostly equivalent. conda vs mamba is a little bit like chasing the latest tooling. It used to be that conda was nearly necessary to handle the packaging metadata from conda-forge, but today, with libmamba-solver, conda should be just as usable as mamba. The problem is that if you try to update an environment with a REALLY OLD conda you might not have libmamba-sovler installed.

2. But what is the use of updating mamba or conda, before updating --all?

sometimes there are bugs in mamba that cause `mamba update --all` to fail.

3. What's the use of that and what are the differences in outcome of updating using the installer?

You really only need the installer the very first time. Once conda is installed, you can simply use `conda update`.

4. What would be the difference in outcome of a fresh reinstall vs. the -u way vs. the mamba update --all way?

Not accumulating packages accidentally. 

5. And what is the preferred way?

I do:
* Keep my base env as clean as possible.
* `mamba update --all --yes` on a weekly basis.


> I also feel it would be great, if the one official way would be mentioned in the docs.

I think we could add 
```
conda update conda
conda update --all
```
to our instructions in the `Installation` section.

PRs welcome to add them to the docs.

**Reply by gernophil** [reply]
Thanks for your reply. That really helped me understanding the differences. Two more questions.
1. Would it matter, if my base is activated before I `mamba update --all` or would it also activate base, if I have no env active?
2. Would `mamba update --all` also update the base Python once in a while or would that always be pinned at the version you installed originally?

**Reply by hmaarrfk** [reply]
1. I'm not sure "not having an environment active" is a supported mode of operation. If things "work" don't count on them continuing to work.

Ensure that the environment you want to update is "activated"


2. When you type `mamba update --all` you will see the output:
```
Pinned packages:
  - python 3.10.*
```

which states that python will not be updated to anything other than `3.10.*`

**Reply by gernophil** [reply]
It seems like `conda update --all` doesn't state the pinned packages. This might be good to include as well :).

**Reply by gernophil** [reply]
Also, how can I be sure Python is still pinned. I don't get any info an that. Also I just did
```
conda update conda
conda update --all
```
and
```
mamba update --all
```
still finds some packages to update.

**Reply by hmaarrfk** [reply]
Updating all packages is a heuristic. 

There may also be some conflicting requirements. 

Because of this you shouldn’t be installing much more than conda and mamba (which come preinstalled) in the base environment

**Reply by gernophil** [reply]
I never install anything into base :).

EDIT: Ok, it only did that on one machine, but not the other. Maybe this was more outdated then the other :).
